In [291]:
import pandas as pd
import numpy as np

In [292]:
df = pd.read_csv('gapminder-unclean.csv')

In [293]:
df.shape

(1704, 7)

In [294]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1704 entries, 0 to 1703
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   country    1704 non-null   object 
 1   continent  1692 non-null   object 
 2   year       1704 non-null   int64  
 3   lifeExp    1703 non-null   float64
 4   pop        1704 non-null   int64  
 5   gdpPercap  1703 non-null   float64
 6   iso_alpha  1704 non-null   object 
dtypes: float64(2), int64(2), object(3)
memory usage: 93.3+ KB


In [295]:
for col in df.columns:
  print(f"{col} Column has {df[col].isnull().sum()} null values")

country Column has 0 null values
continent Column has 12 null values
year Column has 0 null values
lifeExp Column has 1 null values
pop Column has 0 null values
gdpPercap Column has 1 null values
iso_alpha Column has 0 null values


In [296]:
df.isnull().sum().sum()

np.int64(14)

In [297]:
df.head()

country continent  year  lifeExp       pop   gdpPercap iso_alpha
0  Afghanistan      Asia  1952   28.801   8425333  779.445314       AFG
1  Afghanistan      Asia  1957   30.332   9240934  820.853030       AFG
2  Afghanistan      Asia  1962   31.997  10267083  853.100710       AFG
3  Afghanistan      Asia  1967   34.020  11537966  836.197138       AFG
4  Afghanistan      Asia  1972   36.088  13079460  739.981106       AFG

In [298]:
df.tail()

country continent  year  lifeExp       pop   gdpPercap iso_alpha
1699  Zimbabwe    Africa  1987   62.351   9216418  706.157306       ZWE
1700  Zimbabwe    Africa  1992   60.377  10704340  693.420786       ZWE
1701  Zimbabwe    Africa  1997   46.809  11404948  792.449960       ZWE
1702  Zimbabwe    Africa  2002   39.989  11926563  672.038623       ZWE
1703  Zimbabwe    Africa  2007   43.487  12311143  469.709298       ZWE

In [299]:
df.describe()

year      lifeExp           pop      gdpPercap
count  1704.00000  1703.000000  1.704000e+03    1703.000000
mean   1979.50000    59.476932  2.960121e+07    7211.034188
std      17.26533    12.920492  1.061579e+08    9858.756515
min    1952.00000    23.599000  6.001100e+04     241.165876
25%    1965.75000    48.185000  2.793664e+06    1201.919257
50%    1979.50000    60.765000  7.023596e+06    3530.690067
75%    1993.25000    70.846000  1.958522e+07    9319.503534
max    2007.00000    82.603000  1.318683e+09  113523.132900

In [300]:
for col in df.columns:
  print(f"{col} Column has {df[col].nunique()} Unique Values")
  print(f"Its values are {df[col].unique()} " , end = '\n\n\n')

country Column has 142 Unique Values
Its values are ['Afghanistan' 'Albania' 'Algeria' 'Angola' 'Argentina' 'Australia'
 'Austria' 'Bahrain' 'Bangladesh' 'Belgium' 'Benin' 'Bolivia'
 'Bosnia and Herzegovina' 'Botswana' 'Brazil' 'Bulgaria' 'Burkina Faso'
 'Burundi' 'Cambodia' 'Cameroon' 'Canada' 'Central African Republic'
 'Chad' 'Chile' 'China' 'Colombia' 'Comoros' 'Congo, Dem. Rep.'
 'Congo, Rep.' 'Costa Rica' "Cote d'Ivoire" 'Croatia' 'Cuba'
 'Czech Republic' 'Denmark' 'Djibouti' 'Dominican Republic' 'Ecuador'
 'Egypt' 'El Salvador' 'Equatorial Guinea' 'Eritrea' 'Ethiopia' 'Finland'
 'France' 'Gabon' 'Gambia' 'Germany' 'Ghana' 'Greece' 'Guatemala' 'Guinea'
 'Guinea-Bissau' 'Haiti' 'Honduras' 'Hong Kong, China' 'Hungary' 'Iceland'
 'India' 'Indonesia' 'Iran' 'Iraq' 'Ireland' 'Israel' 'Italy' 'Jamaica'
 'Japan' 'Jordan' 'Kenya' 'Korea, Dem. Rep.' 'Korea, Rep.' 'Kuwait'
 'Lebanon' 'Lesotho' 'Liberia' 'Libya' 'Madagascar' 'Malawi' 'Malaysia'
 'Mali' 'Mauritania' 'Mauritius' 'Mexico' 'Mon

- country  : object , has lots of unique values (142) , need encode and may be deleted
- continent : object , small number(5) , need encode
- year   : int , need scale
- lifeExp  : float , its nunique values is close to the df rows (1625) , need scale
- pop : int , its nunique values is the df rows (1704) , need scale , will disappear duplicated rows
- gdpPercap : float , its unique values is really close to the df rows (1703) , need scaling , will disappear duplicated rows
- iso_alpha : object , has lots of unique values (141) , need encode and may be deleted

In [301]:
# handling null values
nulls = ['continent' , 'lifeExp', 'gdpPercap']
for col in nulls:
  if df[col].dtype in ['int64' , 'float64']:
    # null values in numerical data will be filled with median
    df[col].fillna(df[col].median() , inplace=True)
  else:
    # null values in categorical data will be filled with mode
    df[col].fillna(df[col].mode()[0] , inplace=True)

/tmp/ipython-input-1548670511.py:9: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna(df[col].mode()[0] , inplace=True)
/tmp/ipython-input-1548670511.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try us

In [302]:
df.isnull().sum().sum()

np.int64(0)

In [303]:
for col in df.columns:
  print(f"{col} Column has {df[col].isnull().sum()} null values")

country Column has 0 null values
continent Column has 0 null values
year Column has 0 null values
lifeExp Column has 0 null values
pop Column has 0 null values
gdpPercap Column has 0 null values
iso_alpha Column has 0 null values


In [304]:
# handling outliers
numerical_columns = ['year' , 'lifeExp', 'pop', 'gdpPercap']

for col in numerical_columns:
  q1 = df[col].quantile(0.25)
  q3 = df[col].quantile(0.75)
  iqr = q3 - q1
  lower = q1 - 1.5 * iqr
  upper = q3 + 1.5 * iqr
  before = df[(df[col] < lower) | (df[col] > upper)]
  print("Before Handling : ")
  print(f"num of outliers in {col} Column : {len(before)}")
  df[col] = df[col].clip(lower, upper)
  after = df[(df[col] < lower) | (df[col] > upper)]
  print("After Handling : ")
  print(f"num of outliers in {col} Column : {len(after)}" , end = '\n\n')

Before Handling : 
num of outliers in year Column : 0
After Handling : 
num of outliers in year Column : 0

Before Handling : 
num of outliers in lifeExp Column : 0
After Handling : 
num of outliers in lifeExp Column : 0

Before Handling : 
num of outliers in pop Column : 208
After Handling : 
num of outliers in pop Column : 0

Before Handling : 
num of outliers in gdpPercap Column : 143
After Handling : 
num of outliers in gdpPercap Column : 0



In [305]:
from sklearn.preprocessing import MinMaxScaler , StandardScaler

df_copy1 = df.copy()
def column_info(columns , df) :
  for col in columns:
    print(f"Min of {col} Column : {df[col].min()}")
    print(f"Max of {col} Column : {df[col].max()}")
    print(f"Range of {col} Column : {df[col].max() - df[col].min()}")
    print(f"Num of Unique Values of {col} Column : {df[col].nunique()}" , end = '\n\n\n')

normalize_columns = ['lifeExp', 'gdpPercap', 'pop' , 'year']
scaler = MinMaxScaler()
print("Before Normalizing : ")
column_info(normalize_columns , df)

df[normalize_columns] = scaler.fit_transform(df[normalize_columns])

print("After Normalizing")
column_info(normalize_columns , df)

Before Normalizing : 
Min of lifeExp Column : 23.599
Max of lifeExp Column : 82.603
Range of lifeExp Column : 59.00399999999999
Num of Unique Values of lifeExp Column : 1625


Min of gdpPercap Column : 241.1658765
Max of gdpPercap Column : 21488.712491125
Range of gdpPercap Column : 21247.546614625
Num of Unique Values of gdpPercap Column : 1561


Min of pop Column : 60011.0
Max of pop Column : 44772558.375
Range of pop Column : 44712547.375
Num of Unique Values of pop Column : 1497


Min of year Column : 1952
Max of year Column : 2007
Range of year Column : 55
Num of Unique Values of year Column : 12


After Normalizing
Min of lifeExp Column : 0.0
Max of lifeExp Column : 1.0
Range of lifeExp Column : 1.0
Num of Unique Values of lifeExp Column : 1625


Min of gdpPercap Column : 0.0
Max of gdpPercap Column : 0.9999999999999999
Range of gdpPercap Column : 0.9999999999999999
Num of Unique Values of gdpPercap Column : 1561


Min of pop Column : 0.0
Max of pop Column : 1.0
Range of pop Colu

In [306]:
df.describe()

year      lifeExp          pop    gdpPercap
count  1704.000000  1704.000000  1704.000000  1704.000000
mean      0.500000     0.608072     0.300977     0.287695
std       0.313915     0.218913     0.332736     0.314527
min       0.000000     0.000000     0.000000     0.000000
25%       0.250000     0.416904     0.061138     0.045224
50%       0.500000     0.629889     0.155741     0.154819
75%       0.750000     0.800734     0.436683     0.427134
max       1.000000     1.000000     1.000000     1.000000

In [307]:
continent_dummies = pd.get_dummies(df['continent'], prefix='continent')
for col in continent_dummies.columns:
  continent_dummies[col] = continent_dummies[col].astype(int)
df = pd.concat([df, continent_dummies], axis=1)
df = df.drop('continent' , axis = 1)
df

country      year   lifeExp       pop  gdpPercap iso_alpha  \
0     Afghanistan  0.000000  0.088164  0.187091   0.025334       AFG   
1     Afghanistan  0.090909  0.114111  0.205332   0.027283       AFG   
2     Afghanistan  0.181818  0.142329  0.228282   0.028800       AFG   
3     Afghanistan  0.272727  0.176615  0.256705   0.028005       AFG   
4     Afghanistan  0.363636  0.211664  0.291181   0.023476       AFG   
...           ...       ...       ...       ...        ...       ...   
1699     Zimbabwe  0.636364  0.656769  0.204784   0.021884       ZWE   
1700     Zimbabwe  0.727273  0.623314  0.238061   0.021285       ZWE   
1701     Zimbabwe  0.818182  0.393363  0.253730   0.025946       ZWE   
1702     Zimbabwe  0.909091  0.277778  0.265396   0.020279       ZWE   
1703     Zimbabwe  1.000000  0.337062  0.273998   0.010756       ZWE   

      continent_Africa  continent_Americas  continent_Asia  continent_Europe  \
0                    0                   0               1                 0   
1                    0                   0               1                 0   
2                    0                   0               1                 0   
3                    0                   0               1                 0   
4                    0                   0               1                 0   
...                ...                 ...             ...               ...   
1699                 1                   0               0                 0   
1700                 1                   0               0                 0   
1701                 1                   0               0                 0   
1702                 1                   0               0                 0   
1703                 1                   0               0                 0   

      continent_Oceania  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0  
...                 ...  
1699                  0  
1700                  0  
1701                  0  
1702                  0  
1703                  0  

[1704 rows x 11 columns]

In [308]:
from sklearn.preprocessing import LabelEncoder

def column_info2(columns , df):
  for col in columns:
    print(f"For {col} Column")
    print(f"Type : {df[col].dtype}")
    print(f"Unique Values : {df[col].unique()}" , end = '\n\n\n')

object_columns = ['country' , 'iso_alpha']

print("Before Encoding : ")
column_info2(object_columns , df)

label_encoder = LabelEncoder()
for col in object_columns:
  df[col] = label_encoder.fit_transform(df[col])

print("After Encoding : ")
column_info2(object_columns , df)

Before Encoding : 
For country Column
Type : object
Unique Values : ['Afghanistan' 'Albania' 'Algeria' 'Angola' 'Argentina' 'Australia'
 'Austria' 'Bahrain' 'Bangladesh' 'Belgium' 'Benin' 'Bolivia'
 'Bosnia and Herzegovina' 'Botswana' 'Brazil' 'Bulgaria' 'Burkina Faso'
 'Burundi' 'Cambodia' 'Cameroon' 'Canada' 'Central African Republic'
 'Chad' 'Chile' 'China' 'Colombia' 'Comoros' 'Congo, Dem. Rep.'
 'Congo, Rep.' 'Costa Rica' "Cote d'Ivoire" 'Croatia' 'Cuba'
 'Czech Republic' 'Denmark' 'Djibouti' 'Dominican Republic' 'Ecuador'
 'Egypt' 'El Salvador' 'Equatorial Guinea' 'Eritrea' 'Ethiopia' 'Finland'
 'France' 'Gabon' 'Gambia' 'Germany' 'Ghana' 'Greece' 'Guatemala' 'Guinea'
 'Guinea-Bissau' 'Haiti' 'Honduras' 'Hong Kong, China' 'Hungary' 'Iceland'
 'India' 'Indonesia' 'Iran' 'Iraq' 'Ireland' 'Israel' 'Italy' 'Jamaica'
 'Japan' 'Jordan' 'Kenya' 'Korea, Dem. Rep.' 'Korea, Rep.' 'Kuwait'
 'Lebanon' 'Lesotho' 'Liberia' 'Libya' 'Madagascar' 'Malawi' 'Malaysia'
 'Mali' 'Mauritania' 'Mauritiu

that maybe not right but i will keep these columns for now , until we know if they are really useless or not

In [309]:
df.duplicated().sum()

np.int64(0)

In [310]:
df_copy = df.copy()

In [311]:
df_copy = df_copy.drop('gdpPercap' , axis = 1)

In [312]:
df_copy.duplicated().sum()

np.int64(0)

In [313]:
df_copy = df_copy.drop('pop' , axis = 1)

In [314]:
df_copy.duplicated().sum()

np.int64(0)

ok , after all there are no duplicates

In [318]:
print("After Data Cleaning and Preprocessing : ")
print(f"Shape : {df.shape}" , end = '\n\n')
print(f"Column Data Types : {df.dtypes}" , end = '\n\n')
print(f"Null Values : {df.isnull().sum().sum()}" , end = '\n\n')
print(f"Info : {df.info()}" , end = '\n\n')
print(f"Describe : {df.describe()}" , end = '\n\n')
print(f"Duplicates : {df.duplicated().sum()}" , end = '\n\n')

After Data Cleaning and Preprocessing : 
Shape : (1704, 11)

Column Data Types : country                 int64
year                  float64
lifeExp               float64
pop                   float64
gdpPercap             float64
iso_alpha               int64
continent_Africa        int64
continent_Americas      int64
continent_Asia          int64
continent_Europe        int64
continent_Oceania       int64
dtype: object

Null Values : 0

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1704 entries, 0 to 1703
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   country             1704 non-null   int64  
 1   year                1704 non-null   float64
 2   lifeExp             1704 non-null   float64
 3   pop                 1704 non-null   float64
 4   gdpPercap           1704 non-null   float64
 5   iso_alpha           1704 non-null   int64  
 6   continent_Africa    1704 non-null   int64  
 7   cont

In [319]:
excel_filename = 'gapminder_cleaned.xlsx'
df.to_excel(excel_filename, index=False)

csv_filename = 'gapminder_cleaned.csv'
df.to_csv(csv_filename, index=False)